File: framework/headless/feedback.go
package headless

import (
	"fmt"
	"strconv"

	"github.com/DonaldMurillo/gofastr/core-ui/html"
	"github.com/DonaldMurillo/gofastr/core-ui/urlsafe"
	"github.com/DonaldMurillo/gofastr/core/render"
)

// Alert parts.
const (
	PartToneWord Part = "tone-word"
	PartDismiss  Part = "dismiss"
)

// Live says how — and whether — an alert interrupts.
//
// This is the decision the component exists to force, because getting
// it wrong is not a small mistake in either direction.
//
//   - LiveOff is the default and the right answer for anything the
//     server rendered as part of the page. An alert that was already
//     on screen when the page loaded has not "happened"; announcing it
//     as an event means a screen reader reads page furniture over the
//     top of whatever the user was doing, on every single page load.
//   - LivePolite queues the message until the user pauses. It is for
//     something that arrived after load and can wait: a background job
//     finished, a value saved.
//   - LiveAssertive interrupts immediately. It is for something the
//     user must hear before they do anything else: the deploy failed,
//     the form did not submit, the connection dropped. Nothing routine
//     belongs here — an assertive region that fires often is one a
//     user learns to resent and cannot turn off.
//
// role="alert" already implies assertive and role="status" already
// implies polite, so only the role is written. Stating both is how a
// message ends up announced twice.
type Live string

const (
	LiveOff       Live = ""
	LivePolite    Live = "polite"
	LiveAssertive Live = "assertive"
)

// AlertProps is a message about something that happened, or is true.
type AlertProps struct {
	// Tone names the kind of message: the class map turns it into colour
	// through the root's "<part>--<tone>" variant. It is passed
	// through, not interpreted; the tone word is what carries it to a
	// reader.
	Tone string
	// ToneWord is the tone in words — "Error", "Warning", "Success".
	// It is rendered for assistive tech and hidden visually, and it is
	// the reason this component satisfies WCAG 1.4.1: colour may not
	// be the only thing carrying meaning, and a red box is exactly
	// that to anyone who cannot see the red.
	//
	// Empty means the Title already says which kind it is — "Deploy
	// failed" needs no "Error:" in front of it — so the word is
	// skipped rather than duplicated.
	ToneWord string
	// Title is the headline. Required: an alert with no headline is a
	// coloured paragraph.
	Title string
	// Text is the detail, in prose.
	Text string
	// Icon is decorative. The tone word carries the meaning.
	Icon render.HTML
	// Actions are the controls: retry, view logs, dismiss.
	Actions render.HTML
	// Live says whether this interrupts. See Live.
	Live Live
	// Focus moves focus here when the page loads.
	//
	// It is how a message that is ALREADY THERE gets announced. A live
	// region announces changes; content present when the document
	// loads is not a change, and role="alert" at load time is reported
	// inconsistently across screen readers. After a form posts and the
	// server redirects — the shape of every action in a
	// server-rendered app — the confirmation is present at load, so
	// the only reliable way to say it is to put the reader on it.
	Focus bool
	// DismissHref makes the alert dismissable with a link that keeps a
	// real href, so dismissing needs no script and survives the page
	// being reloaded. The label names what is being dismissed, because
	// "Dismiss" three times in a row tells a screen reader user
	// nothing.
	DismissHref  string
	DismissLabel string
	// Island is where the dismiss goes with script: dismissing is an
	// in-page state change, so the × carries the RPC contract beside
	// its href. Required when DismissHref is set; ignored otherwise.
	Island Island

	ID         string
	ExtraAttrs html.Attrs

	// Parts: attrs and binds on every part. Nothing here is
	// fillable — Actions already takes the page's own controls, and
	// everything else an alert draws is what a screen reader is given
	// to tell one alert from another.
	Parts Parts
	// Strings are the strings this component says. Nil means the English
	// defaults; a layer above sets them from the request's language.
	Strings *Strings
}

// Alert renders the message.
func Alert(p AlertProps, s Classes) render.HTML {
	b := p.Parts.Box(s)
	if p.Title == "" {
		panic("headless: Alert requires Title")
	}
	own := Merge(Safe(p.ExtraAttrs, "role"), Attrs(map[string]string{"id": p.ID}))
	if p.Focus {
		// Focusable without joining the tab order: focus is put here
		// once, and Tab from it carries on into the page.
		own["tabindex"] = "-1"
		Mark(own, "autofocus")
	}
	switch p.Live {
	case LiveAssertive:
		own["role"] = "alert"
	case LivePolite:
		own["role"] = "status"
	}
	if cls := s.Variant(PartRoot, p.Tone); cls != "" {
		own["class"] = cls
	}

	head := make([]render.HTML, 0, 3)
	if p.Icon != "" {
		head = append(head, b.El("span", PartIcon,
			Attrs(map[string]string{"aria-hidden": "true"}), p.Icon))
	}
	title := make([]render.HTML, 0, 2)
	if p.ToneWord != "" {
		// Read before the title, so the kind of message arrives before
		// the message. The trailing space is inside the hidden span:
		// without it screen readers run the two together ("ErrorDeploy
		// failed").
		title = append(title, b.El("span", PartToneWord, nil, render.Text(p.ToneWord+": ")))
	}
	title = append(title, render.Text(p.Title))
	head = append(head, b.El("p", PartTitle, nil, title...))

	kids := []render.HTML{b.El("div", PartHeader, nil, head...)}
	if p.Text != "" {
		kids = append(kids, b.El("p", PartDesc, nil, render.Text(p.Text)))
	}
	if p.Actions != "" {
		kids = append(kids, b.El("div", PartFooter, nil, p.Actions))
	}
	if p.DismissHref != "" {
		requireIsland("Alert with DismissHref", p.Island)
		if urlsafe.CleanAnchor(p.DismissHref) == "" {
			panic("headless: Alert DismissHref " + strconv.Quote(p.DismissHref) + " is not a URL the anchor policy allows")
		}
		label := p.DismissLabel
		if label == "" {
			label = fmt.Sprintf(p.Strings.Resolve().DismissTitled, p.Title)
		}
		// The same element is both destinations: the href is the page
		// without script, the island contract is the region update
		// with it.
		dismiss := Merge(Attrs(map[string]string{"href": p.DismissHref, "aria-label": label}),
			p.Island.attrs(p.DismissHref, "GET"))
		kids = append(kids, b.El("a", PartDismiss, dismiss, render.Text("×")))
	}
	return b.El("div", PartRoot, own, kids...)
}

func init() {
	Register(Spec{
		Name:    "Alert",
		Anatomy: []Part{PartRoot, PartHeader, PartIcon, PartToneWord, PartTitle, PartDesc, PartFooter, PartDismiss},
		WithParts: func(s Classes, parts Parts) render.HTML {
			return Alert(AlertProps{Title: "Deploy failed", Text: "Exit 1 in the test stage.", Tone: "danger",
				ToneWord: "Error", Icon: SpecimenGlyph, Actions: render.HTML("<a href=\"/logs\">View logs</a>"),
				DismissHref: "/apps?dismiss=1", Island: Island{Endpoint: "/island/alerts", Signal: "alerts"},
				Parts: parts}, s)
		},
		Cases: func(k Kit) []Case {
			s := k.Classes
			return []Case{{
				Name: "loud",
				Why:  "something went wrong and the reader must be interrupted: assertive, with the tone said in words as well as drawn in colour",
				HTML: Alert(AlertProps{
					Title: "Deploy failed", Text: "The build exited 1 after 42 seconds.",
					ToneWord: "Error", Live: LiveAssertive, Icon: SpecimenGlyph,
					Actions:     Button(ButtonProps{Label: "View logs", Variant: "secondary"}, k.For("Button")),
					DismissHref: "/apps?dismiss=1", DismissLabel: "Dismiss deploy failure",
					Island: Island{Endpoint: "/island/alerts", Signal: "alerts"},
				}, k.Variant("Alert", "danger")),
			}, {
				Name: "quiet",
				Why:  "the default: a standing message that is on screen anyway and must not interrupt whatever is being read",
				HTML: Alert(AlertProps{Title: "3 apps need attention"}, s),
			}, {
				Name: "already there",
				Why:  "the confirmation after a post-and-redirect — present at load, so the only reliable way to announce it is to put the reader on it",
				HTML: Alert(AlertProps{Title: "App restarted", Focus: true, ID: "restarted"},
					k.Variant("Alert", "success")),
			}}
		},
	})

	Register(Spec{
		Name:    "Button",
		Anatomy: []Part{PartRoot, PartIcon},
		Cases: func(k Kit) []Case {
			s := k.Classes
			return []Case{{
				Name: "labelled",
				Why:  "the ordinary case, and the reason type is always stated: inside a form the default is submit",
				HTML: Button(ButtonProps{Label: "Save changes", Variant: "primary"}, s),
			}, {
				Name: "icon only",
				Why:  "a button with no text still needs a name, or it cannot be operated by anyone who cannot see it",
				HTML: Button(ButtonProps{AriaLabel: "Close", Icon: SpecimenGlyph, Variant: "ghost"}, s),
			}, {
				Name: "link that looks like a button",
				Why:  "it navigates, so it is an anchor — a button that changes the URL is a button a middle click cannot open",
				HTML: Button(ButtonProps{Label: "Read the docs", Variant: "primary", Href: "/docs"}, s),
			}, {
				Name: "disabled link",
				Why:  "a disabled anchor is not a thing in HTML, so the href goes and aria-disabled says why, rather than leaving a live link that looks dead",
				HTML: Button(ButtonProps{Label: "Rollback", Variant: "danger", Href: "/rollback", Disabled: true}, s),
			}, {
				Name: "firing a request",
				Why: "the request a click makes is a prop on the button that fires it, in one reviewable place: " +
					"the rpc triple lands through the Action seam and cannot arrive any other way, so a button " +
					"that fires a request says so in its props",
				HTML: Button(ButtonProps{
					Label: "Restart api", Variant: "secondary",
					Action: html.Attrs{
						"data-fui-rpc":        "/apps/api/restart",
						"data-fui-rpc-method": "POST",
						"data-fui-rpc-signal": "apps",
					},
				}, s),
			}, {
				Name: "mutating a local signal",
				Why: "a click that never leaves the browser is still what the button DOES, so it travels the same " +
					"seam as a request: the mutation's attribute sits on the button beside its label, where a " +
					"reviewer reads them together",
				HTML: Button(ButtonProps{
					Label: "Add replica", Variant: "ghost",
					Action: html.Attrs{"data-fui-signal-inc": "replicas"},
				}, s),
			}}
		},
	})
}

File: framework/headless/feedback.js
// headless-feedback: the behaviour module for the feedback family —
// copy controls, the toast stack runtime, the notification bell's
// spoken count, and the offline banner's retry control. Loaded by the
// kernel on one of its markers, at boot, on insertion, or after a
// client navigation.
//
// The toast runtime is the API the kernel's response-header path and
// every window.__gofastr.toast caller already speak (NS.toast,
// NS._initToasts, NS._dismissToast, NS._toastTimers, NS._toastSeq):
// it moved here from the retired core-ui/runtime "toasts" module, and
// the kernel's loadModule('headless-feedback') retarget is what keeps
// the X-Gofastr-Toast dispatch working. Every sentence the module
// writes arrived as a data-hui-* attribute the component rendered
// from its Strings — the dismiss label and the copy status travel on
// the stack and the copy control — so a translated page announces in
// its own language.
(function () {
  'use strict';
  const NAME = 'headless-feedback';
  const NS = window.__gofastr = window.__gofastr || {};
  if (NS.loadedModules && Object.prototype.hasOwnProperty.call(NS.loadedModules, NAME)) return;
  NS.loadedModules = NS.loadedModules || {};
  NS.loadedModules[NAME] = true;

  function within(root, sel) {
    const out = [];
    if (root.matches && root.matches(sel)) out.push(root);
    if (root.querySelectorAll) out.push.apply(out, root.querySelectorAll(sel));
    return out;
  }

  // ─── toast stack runtime (moved from src/toasts.js) ─────────────

  NS._toastTimers = NS._toastTimers || new Map();
  NS._toastSeq = NS._toastSeq || 0;

  NS._initToasts = function (root) {
    // A row the module built carries its id; a row the server
    // rendered carries the toast marker and, when it has a lifetime,
    // the TTL. Both arm here; a server row is given its id on sight.
    const items = root.querySelectorAll('[data-hui-toast-id], [data-hui-toast-ttl-ms]');
    const present = new Set();
    items.forEach(function (item) {
      if (!item.hasAttribute('data-hui-toast-id')) item.setAttribute('data-hui-toast-id', 's' + (++NS._toastSeq));
      const id = item.getAttribute('data-hui-toast-id');
      present.add(id);
      // Ids are attribute-borne, so the registry is a Map: plain
      // string keys, no prototype re-parenting.
      if (NS._toastTimers.get(id)) return;
      const ttl = parseInt(item.getAttribute('data-hui-toast-ttl-ms') || '0', 10);
      if (ttl > 0) {
        const rec = { remaining: ttl, startedAt: Date.now(), timer: 0 };
        NS._toastTimers.set(id, rec);
        const arm = function () {
          rec.startedAt = Date.now();
          rec.timer = setTimeout(function () { NS._dismissToast(item, id); }, rec.remaining);
        };
        const pause = function () {
          if (!rec.timer) return;
          clearTimeout(rec.timer);
          rec.timer = 0;
          rec.remaining -= Date.now() - rec.startedAt;
          if (rec.remaining < 100) rec.remaining = 100;
        };
        item.addEventListener('mouseenter', pause);
        item.addEventListener('focusin', pause);
        item.addEventListener('mouseleave', arm);
        item.addEventListener('focusout', arm);
        arm();
      }
      // A server row's dismiss is a link with its Island: the kernel
      // owns that click. The module's own rows dismiss here.
      if (item.hasAttribute('data-hui-toast')) return;
      item.addEventListener('click', function (e) {
        if (e.target.closest('[data-hui-toast-dismiss]')) {
          e.preventDefault();
          NS._dismissToast(item, id);
        }
      });
    });
    for (const entry of Array.from(NS._toastTimers)) {
      if (!present.has(entry[0])) {
        clearTimeout(entry[1].timer);
        NS._toastTimers.delete(entry[0]);
      }
    }
  };

  NS._dismissToast = function (item, id) {
    if (!item || item.classList.contains('is-leaving')) return;
    item.classList.add('is-leaving');
    const rec = NS._toastTimers.get(id);
    if (rec) { clearTimeout(rec.timer); NS._toastTimers.delete(id); }
    const cs = getComputedStyle(item);
    const ms = parseFloat(cs.animationDuration) * 1000 || 200;
    setTimeout(function () { if (item.parentNode) item.parentNode.removeChild(item); }, ms);
  };

  // stackLabelFmt: the dismiss label template the stack carries (from
  // Strings.DismissTitled), %s where the toast's title goes. The
  // module never says the sentence itself.
  function stackLabelFmt(container) {
    return (container && container.getAttribute('data-hui-toast-dismiss-label')) || '%s';
  }

  NS.toast = function (cfg) {
    if (cfg == null) return null;
    if (typeof cfg === 'string') cfg = { title: cfg, ttl: 4000 };
    if (!cfg.title) return null;
    let container = null;
    if (cfg.stack) {
      container = document.querySelector('[data-fui-toast-stack="' + CSS.escape(cfg.stack) + '"]');
    }
    if (!container) container = document.querySelector('[data-fui-toast-stack]');
    if (!container) {
      // Body singleton, created at most once and re-attached by the
      // SPA full-shell swap — distinct from the kernel's unstyled
      // fallback container for the module-failed-to-load path.
      container = NS.doc.singleton('fui-toast-stack-auto', function () {
        const c = document.createElement('div');
        c.className = 'fui-toast-stack';
        c.setAttribute('data-fui-comp', 'ui-toast-stack');
        c.setAttribute('data-fui-toast-stack', '__auto');
        c.style.cssText = 'position:fixed;top:1rem;right:1rem;z-index:2147483600;display:grid;gap:0.5rem;pointer-events:none;max-width:min(360px,calc(100vw - 2rem));';
        return c;
      });
      if (NS.scanAndLoadCSS) NS.scanAndLoadCSS(container);
    }

    const id = 't' + (++NS._toastSeq);
    const variant = cfg.variant || 'info';
    const assertive = variant === 'warning' || variant === 'danger';
    const glyph = ({ success: '✓', warning: '!', danger: '✕', info: '•', neutral: '•' })[variant] || '•';

    const item = document.createElement('div');
    item.className = 'fui-toast-stack__item';
    item.setAttribute('data-hui-toast-id', id);
    const ttl = parseInt(cfg.ttl || 0, 10);
    if (ttl > 0) item.setAttribute('data-hui-toast-ttl-ms', String(ttl));

    const wrap = document.createElement('div');
    wrap.className = 'fui-notification fui-notification--' + variant;
    wrap.setAttribute('data-fui-comp', 'ui-notification');
    wrap.setAttribute('data-hui-toast', '');
    wrap.setAttribute('role', assertive ? 'alert' : 'status');
    wrap.setAttribute('aria-live', assertive ? 'assertive' : 'polite');

    // The tone word, read and not shown, from the stack's own Strings:
    // the module says no word of its own.
    const toneAttr = {
      info: 'data-hui-toast-tone-info', success: 'data-hui-toast-tone-success',
      warning: 'data-hui-toast-tone-warning', danger: 'data-hui-toast-tone-danger'
    }[variant];
    const toneWord = toneAttr ? container.getAttribute(toneAttr) : '';
    let toneEl = null;
    if (toneWord) {
      toneEl = document.createElement('span');
      toneEl.className = 'fui-visually-hidden';
      toneEl.textContent = toneWord + ': ';
    }
    const icon = document.createElement('span');
    icon.className = 'fui-notification__icon';
    icon.setAttribute('aria-hidden', 'true');
    icon.textContent = glyph;

    const titleEl = document.createElement('span');
    titleEl.className = 'fui-notification__title';
    titleEl.textContent = cfg.title;
    let bodyEl = null;
    if (cfg.body) {
      bodyEl = document.createElement('span');
      bodyEl.className = 'fui-notification__body';
      bodyEl.textContent = cfg.body;
    }

    const dismiss = document.createElement('button');
    dismiss.type = 'button';
    dismiss.className = 'fui-notification__dismiss';
    dismiss.setAttribute('aria-label', stackLabelFmt(container).replace('%s', cfg.title));
    dismiss.setAttribute('data-hui-toast-dismiss', '');
    dismiss.textContent = '×';

    if (toneEl) wrap.appendChild(toneEl);
    wrap.appendChild(icon);
    wrap.appendChild(titleEl);
    if (bodyEl) wrap.appendChild(bodyEl);
    wrap.appendChild(dismiss);
    item.appendChild(wrap);
    container.appendChild(item);

    // The stack's capacity: past it the oldest row goes, so a burst
    // of toasts cannot pile a column over the page.
    const max = parseInt(container.getAttribute('data-hui-toast-max') || '0', 10);
    if (max > 0) {
      const rows = container.querySelectorAll('[data-hui-toast-id]');
      while (rows.length > max) container.removeChild(rows[0]);
    }

    if (NS.scanAndLoadCSS) NS.scanAndLoadCSS(item);
    NS._initToasts(container);
    return id;
  };

  // ─── copy ────────────────────────────────────────────────────────

  // The target stays readable and selectable; without script the page
  // promises nothing about the clipboard. The words (button label,
  // copied label, status sentence with {name}) travel on the wrapper
  // from Strings.
  document.addEventListener('click', function (e) {
    const t = e.target;
    const wrap = t && t.closest && t.closest('[data-hui-copy]');
    if (!wrap) return;
    const id = wrap.getAttribute('data-hui-copy-target');
    const target = id ? document.getElementById(id) : null;
    if (!target) return;
    const text = (target.innerText || target.textContent || '').trim();
    if (!text) return;
    if (navigator.clipboard && navigator.clipboard.writeText) {
      navigator.clipboard.writeText(text).catch(function () {});
    }
    wrap.setAttribute('data-hui-copy-state', 'done');
    const copyBtn = wrap.querySelector('button');
    if (copyBtn) copyBtn.classList.add('fui-copied');
    const btn = wrap.querySelector('[data-hui-copy-label]');
    const fmts = wrap.getAttribute('data-hui-copy-copied') || '';
    if (btn && fmts) btn.textContent = fmts;
    const status = wrap.querySelector('[data-hui-copy-status]');
    if (status) {
      const name = wrap.getAttribute('data-hui-copy-name') || id || '';
      const sentence = (wrap.getAttribute('data-hui-copy-sentence') || '').replace('{name}', name);
      status.textContent = '';
      requestAnimationFrame(function () { status.textContent = sentence; });
    }
    const back = wrap.getAttribute('data-hui-copy-back') || '';
    setTimeout(function () {
      wrap.removeAttribute('data-hui-copy-state');
      if (copyBtn) copyBtn.classList.remove('fui-copied');
      if (btn && back) btn.textContent = back;
    }, 1200);
    // A toast on copy rides this module's own toast runtime; the
    // config JSON travels on the button.
    const toastCfg = (btn && btn.getAttribute('data-hui-copy-toast')) || '';
    if (toastCfg) {
      try { NS.toast(JSON.parse(toastCfg)); } catch (_) {}
    }
  });

  // ─── notification bell ───────────────────────────────────────────

  // The spoken count follows the signal the badge follows: the kernel
  // writes the number into the bound nodes, and the module re-formats
  // the anchor's accessible name through the sentence shape the
  // component rendered.
  document.addEventListener('gofastr:signal', function (e) {
    const d = e && e.detail;
    if (!d || typeof d.name !== 'string') return;
    for (const bell of document.querySelectorAll('[data-hui-notification-bell]')) {
      if (bell.getAttribute('data-fui-signal') !== d.name) continue;
      const fmt = bell.getAttribute('data-hui-notification-count-fmt') || '';
      const n = parseInt(d.value, 10);
      if (!Number.isFinite(n)) return;
      if (fmt) {
        bell.setAttribute('aria-label', fmt.replace('%d', String(n)).replace('%d', String(n)));
      }
      // The badge's count attribute follows the signal too, so the
      // next reader of it (a stylesheet's 99+ shaping, a test) sees
      // the same number the anchor says.
      const badge = bell.querySelector('[data-hui-notification-count]');
      if (badge) badge.setAttribute('data-hui-notification-count', String(n));
    }
  });

  // ─── network retry ───────────────────────────────────────────────

  // The banner itself is the headless module's offline SystemBanner;
  // this half owns the retry: the anchor is a real link (no script =
  // the page reloads through it), and with script the fetch happens
  // in place — a 2xx reports recovery and hides the banner, a failed
  // probe leaves it shown.
  document.addEventListener('click', function (e) {
    const t = e.target;
    const retry = t && t.closest && t.closest('[data-hui-network-retry]');
    if (!retry) return;
    e.preventDefault();
    const href = retry.getAttribute('href') || '';
    if (!href) return;
    // A 2xx from the health endpoint is the reconnect: hide the
    // banner (reportRecovery drives every mounted offline one). A
    // failed probe leaves it shown — the connection is still down
    // and hiding it would lie.
    fetch(href, { credentials: 'same-origin' })
      .then(function (r) { if (r.ok) NS.networkStatus.reportRecovery(); })
      .catch(function () {});
  });

  // The public status API app code called on the retired module,
  // kept on the namespace so callers do not break: reportFailure and
  // reportRecovery drive every mounted offline banner's data-state.
  NS.networkStatus = NS.networkStatus || {};
  NS.networkStatus.reportFailure = function () {
    for (const b of document.querySelectorAll('[data-hui-system-offline]')) b.hidden = false;
  };
  NS.networkStatus.reportRecovery = function () {
    for (const b of document.querySelectorAll('[data-hui-system-offline]')) b.hidden = true;
  };

  // ─── the arrival pass ────────────────────────────────────────────

  function scan(root) {
    const scope = root && root.querySelectorAll ? root : document;
    for (const c of within(scope, '[data-hui-toast-stack],[data-fui-toast-stack]')) NS._initToasts(c);
  }

  scan(document);
  NS._moduleScanners = NS._moduleScanners || {};
  NS._moduleScanners[NAME] = scan;
})();
